iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java

```diff
@@ -21,31 +21,44 @@ class MqttDirectMethod extends Mqtt
 {
     private final String subscribeTopic;
     private final String responseTopic;
-    private boolean isStarted = false;
-
-    private static final String POUND = "#";
-    private static final String BACKSLASH = "/";
-    private static final String QUESTION = "?";
+    private final boolean isConnectingToMqttGateway;
+    private final String deviceId;
 
-    private static final String METHOD = "$iothub/methods/";
-    private static final String POST = METHOD + "POST";
-    private static final String RES = METHOD + "res";
-    private static final String REQ_ID = QUESTION + "$rid=";
+    private boolean isStarted = false;
 
-    //Placement for $iothub/methods/POST/{method name}/?$rid={request id}
-    private static final int METHOD_TOKEN = 3;
-    private static final int REQID_TOKEN = 4;
+    private static final String POST_HUB = "$iothub/methods/POST";
+    private static final String POST_E4K = "$iothub/methods/%s/POST";
+    private static final String RES_HUB = "$iothub/methods/res";
+    private static final String RES_E4K = "$iothub/methods/%s/res";
 
     public MqttDirectMethod(
         String deviceId,
         MqttConnectOptions connectOptions,
         Map<Integer, Message> unacknowledgedSentMessages,
-        Queue<Pair<String, MqttMessage>> receivedMessages)
+        Queue<Pair<String, MqttMessage>> receivedMessages,
+        boolean isConnectingToMqttGateway)
     {
         super(null, deviceId, connectOptions, unacknowledgedSentMessages, receivedMessages);
 
-        this.subscribeTopic = POST + BACKSLASH + POUND;
-        this.responseTopic = RES;
+        this.isConnectingToMqttGateway = isConnectingToMqttGateway;
+        this.deviceId = deviceId;
+
+        if (!isConnectingToMqttGateway)
+        {
+            // $iothub/methods/POST/#
+            this.subscribeTopic = POST_HUB + "/#";
+
+            // $iothub/methods/res
+            this.responseTopic = RES_HUB;
+        }
+        else
+        {
+            // $iothub/methods/{clientId}/POST/#
+            this.subscribeTopic = String.format(POST_E4K, deviceId) + "/#";
```

Comment:
Should ```deviceId``` by ```deviceId/modules/moduleId``` for module cases here?

Reply:
Actually, it looks like we weren't doing that before, either. Odd. I'll take a look at why our existing SDK doesn't seem to subscribe to module-specific methods here.

Reply:
clientId is just `{deviceId}/{moduleId}`.

Reply:
> clientId is just `{deviceId}/{moduleId}`.

I just realized that MqttMessaging in our existing SDK is using "deviceId/modules/moduleId" as "[clientId](https://github.com/Azure/azure-iot-sdk-java/blob/main/iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttMessaging.java#L52)" in topics, so we may also need to replace this with "deviceId/moduleId" for E4K topics, right?

Reply:
Yes. And when subscribing to direct methods, we will need to include this deviceId/moduleId as the clientId. In our existing SDK, it is fine that the subscription topic is the same for devices and modules